Libraries/LibJS/Runtime/PropertyName.h

```diff
@@ -36,41 +37,54 @@ class PropertyName {
         Invalid,
         Number,
         String,
+        Symbol,
     };
 
-    PropertyName() {}
+    PropertyName();
 
-    explicit PropertyName(i32 index)
-        : m_type(Type::Number)
-        , m_number(index)
-    {
-        ASSERT(m_number >= 0);
-    }
+    PropertyName(i32 index);
 
-    explicit PropertyName(const FlyString& string)
-        : m_type(Type::String)
-        , m_string(string)
-    {
-    }
+    PropertyName(const char* string);
 
-    bool is_valid() const { return m_type != Type::Invalid; }
-    bool is_number() const { return m_type == Type::Number; }
-    bool is_string() const { return m_type == Type::String; }
+    PropertyName(const String& string);
 
-    i32 as_number() const { return m_number; }
-    const FlyString& as_string() const { return m_string; }
+    PropertyName(const FlyString& string);
 
-    String to_string() const
-    {
-        if (is_string())
-            return as_string();
-        return String::number(as_number());
-    }
+    PropertyName(const Symbol& symbol);
+
+    bool is_valid() const;
+    bool is_number() const;
+    bool is_string() const;
+    bool is_symbol() const;
+
+    i32 as_number() const;
+    const FlyString& as_string() const;
+    const Symbol& as_symbol() const;
+
+    String to_string() const;
+    u32 hash() const;
 
 private:
     Type m_type { Type::Invalid };
     FlyString m_string;
     i32 m_number { 0 };
+    const Symbol* m_symbol { nullptr };
 };
 
+bool operator==(const PropertyName& lhs, const PropertyName& rhs);
+
 }
+
+template<>
+struct AK::Traits<JS::PropertyName> : public GenericTraits<JS::PropertyName> {
+    static unsigned hash(const JS::PropertyName& key)
+    {
+        ASSERT(key.is_valid());
+        if (key.is_string())
+            return key.as_string().hash();
+        if (key.is_number())
+            return key.as_number();
+        // FIXME: Better symbol hashing
+        return pair_int_hash(key.as_symbol().description().hash(), reinterpret_cast<intptr_t>(&key.as_symbol()));
```

Comment:
This is making the build fail right now. I don't know a good way to hash Symbols because they don't have any data members that make them unique.

Reply:
You can use `ptr_hash()` on pointers instead of casting to `intptr_t` which may confuse the host-side build.